### DISCUSSION TITLE
pyhf yield table script

### DISCUSSION THREAD

**Discussion by Daniel-Noel** [scikit-hep/pyhf#1187] [Ideas]
Hi,

I attach a script I use for writing out pyhf yield tables to a pdf [pyhf_utils.txt](https://github.com/scikit-hep/pyhf/files/5568433/pyhf_utils.txt).

This uses the fitted parameters and errors from pyhf.infer.mle.fit(
        data, model, return_uncertainties=True
    )  as recommended by @kratsg.

I realise this most likely needs some tidying up, but hopefully this is a start and will be useful to share.

Cheers,
Daniel

**Comment by matthewfeickert** [comment]
Hi @Daniel-Noel, maybe instead of text files you could make a [Gist](https://gist.github.com/) and just link it here instead?

**Comment by Daniel-Noel** [comment]
Sure, the file can be found here https://gist.github.com/Daniel-Noel/c4307b7a4355c99240d4dbc6a02913b1

**Reply by lukasheinrich** [reply]
Thanks @Daniel-Noel -- have you considered using tabulate for the yield tables? this way you can print out tables not only in latex , but also ascii etc (and it will do a lot of the formatting for you)

https://pypi.org/project/tabulate/

**Reply by Daniel-Noel** [reply]
Hi @lukasheinrich, thanks for the recommendation - I've had a look at using tabulate for the yield tables in https://gist.github.com/Daniel-Noel/b4685a0053c2ac8ee9ffafce20781592 - it seems to have tidied it up quite a bit. I've also added in rounding to the pdg recommendations

**Comment by alexander-held** [comment]
Hi, great to see this work! I gave the gist a try and it ran fine for me. I did not quite understand the implementation of the yield uncertainty calculation, and I think it is not quite correct. [Here is an example](https://github.com/alexander-held/cabinetry/blob/becf99449538d5a07b4d10cc347090b392c07942/src/cabinetry/model_utils.py#L176-L263) for calculating the total model uncertainty per bin (pre- and post-fit), for the post-fit calculation the parameter correlations are needed. A similar calculation can be done to get the per-sample uncertainties as well.

**Reply by lukasheinrich** [reply]
is this something we can integrate either into `pyhf.contrib` of `cabinetry`?

**Reply by alexander-held** [reply]
Do you mean the yield table creation or the uncertainty calculation? The uncertainty calculation is in `cabinetry`, but it does not seem unreasonable to move it somewhere else. The yield table functionality is needed for analyzers, so needs to be available through `cabinetry` as well. I am happy to use an implementation available somewhere else, though for this specific use case I am not sure whether a nice interface can be created that allows sufficient customization over the formatting (which may be needed for analyzers).

**Reply by alexander-held** [reply]
Maybe one way to factor things is to split out a function that takes (workspace, fit results) and produces (total yield and uncertainty, per-sample yield and uncertainty) for all channels and returns that. There could then be multiple implementations that turn that output into a table, since that part can come down to preference.

**Reply by alexander-held** [reply]
Update: basic yield table functionality now exists in `cabinetry` https://github.com/alexander-held/cabinetry/pull/159, but the interface should be improved a bit. The uncertainty calculations are somewhat expensive for complex models, so the results should be re-used as much as possible when they are needed (e.g. for plots and tables).

**Comment by alexander-held** [comment]
Hi, I put together an example that also performs uncertainty calculation via bootstrapping in a different context: https://github.com/scikit-hep/pyhf/issues/1359#issuecomment-867638285. Maybe this can be useful. See also @HDembinski's nice [tutorial](https://nbviewer.jupyter.org/github/scikit-hep/iminuit/blob/master/tutorial/how_to_draw_error_bands.ipynb) for more information. The error propagation method via `iminuit` described in the tutorial can also be applied to `pyhf`, and is similar (though not exactly equivalent) to the [`cabinetry` implementation](https://github.com/alexander-held/cabinetry/blob/d888ec054e080f474a12de6de375857631c55987/src/cabinetry/model_utils.py#L176-L281). The methods use various approximations and will in general not give the exact same results.
